pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -54,230 +56,196 @@ func (e transientError) Error() string {
-}
-
-func (i *deploymentClientImpl) listDeploymentsForConfig(namespace, configName string) (*kapi.ReplicationControllerList, error) {
-	return i.listDeploymentsForConfigFunc(namespace, configName)
-}
-
-func (i *deploymentClientImpl) updateDeployment(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error) {
-	return i.updateDeploymentFunc(namespace, deployment)
-}
-
-// findDetails inspects the given deployment configuration for any failure causes
-// and returns any found details about it
-// TODO(danmace): Don't use findDetails yet. See: https://github.com/openshift/origin/pull/5530
-func (c *DeploymentConfigController) findDetails(config *deployapi.DeploymentConfig) (string, *kapi.ReplicationControllerList, bool, error) {
-	// Check if any existing inflight deployments (any non-terminal state).
-	existingDeployments, err := c.deploymentClient.listDeploymentsForConfig(config.Namespace, config.Name)
-	if err != nil {
-		return "", nil, false, fmt.Errorf("couldn't list deployments for deployment config %q: %v", deployutil.LabelForDeploymentConfig(config), err)
-	}
-	// check if the latest deployment exists
-	// we'll return after we've dealt with the multiple-active-deployments case
-	latestDeploymentExists, latestDeploymentStatus := deployutil.LatestDeploymentInfo(config, existingDeployments)
-	if latestDeploymentExists && latestDeploymentStatus != deployapi.DeploymentStatusFailed {
-		// If the latest deployment exists and is not failed, clear the dc message
-		return "", existingDeployments, latestDeploymentExists, nil
-	}
-
-	// Rest of the code will handle non-existing or failed latest deployment causes
-	// TODO: Inspect pod logs in case of failed latest
-
-	details := ""
-	allDetails := []string{}
-	if config.Details != nil && len(config.Details.Message) > 0 {
-		// Populate details with the previous message so that in case we stumble upon
-		// an unexpected client error, the message won't be overwritten
-		details = config.Details.Message
+// reconcileDeployments reconciles existing deployment replica counts which
+// could have diverged outside the deployment process (e.g. due to auto or
+// manual scaling, or partial deployments). The active deployment is the last
+// successful deployment, not necessarily the latest in terms of the config
+// version. The active deployment replica count should follow the config, and
+// all other deployments should be scaled to zero.
+//
+// Previously, scaling behavior was that the config replica count was used
+// only for initial deployments and the active deployment had to be scaled up
+// directly. To continue supporting that old behavior we must detect when the
+// deployment has been directly manipulated, and if so, preserve the directly
+// updated value and sync the config with the deployment.
+func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *kapi.ReplicationControllerList, config *deployapi.DeploymentConfig) error {
```

Comment:
I got lost in this method.  I won't hold this up on it if everyone else has read and understood, but I think documenting a state machine would help.